mesa/experimental/jupyter_viz.py

```diff
@@ -37,32 +40,58 @@ def JupyterViz(
         play_interval: play interval (default: 400)
     """
 
-    current_step, set_current_step = solara.use_state(0)
-
     # 1. Set up model parameters
     user_params, fixed_params = split_model_params(model_params)
     model_parameters, set_model_parameters = solara.use_state(
         {**fixed_params, **{k: v["value"] for k, v in user_params.items()}}
     )
 
+    model, set_model = solara.use_state(model_class(**model_parameters))
+    model_cache, set_model_cache = solara.use_state({0: model})
+
     # 2. Set up Model
     def make_model():
         model = model_class(**model_parameters)
-        set_current_step(0)
-        return model
+        set_model(model)
+        set_model_cache({0: model})
 
-    reset_counter = solara.use_reactive(0)
-    model = solara.use_memo(
-        make_model, dependencies=[*list(model_parameters.values()), reset_counter.value]
+    solara.use_memo(
+        make_model,
+        dependencies=list(model_parameters.values()),
     )
 
     def handle_change_model_params(name: str, value: any):
         set_model_parameters({**model_parameters, name: value})
 
+    def handle_step(step: Optional[int] = None):
+        if not model.running:
+            return
+
+        if step in model_cache:
+            updated_model = model_cache[step]
+        else:
+            updated_model = copy.deepcopy(model)
+            updated_model.step()
+
+        if step is not None:
+            set_model_cache({**model_cache, step: updated_model})
+
+        set_model(updated_model)
+
     # 3. Set up UI
     solara.Markdown(name)
     UserInputs(user_params, on_change=handle_change_model_params)
-    ModelController(model, play_interval, current_step, set_current_step, reset_counter)
+    TimelineControls(
+        play_interval=play_interval,
+        on_step=handle_step,
+        on_reset=make_model,
+        current_step=model.schedule.steps,
+        max_step=max(model_cache.keys()),
+    ) if timeline else BaseControls(
+        play_interval=play_interval,
+        on_step=handle_step,
```

Comment:
`handle_step` does caching regardless of whether the control is `TimelineControls` or `BaseControls`. It's not a problem for models that can be visualized, but it's still consuming more RAM than necessary.

Reply:
Clarification: *It's not a problem for the **scale** of models that can be visualized, which is usually small.

Reply:
> but it's still consuming more RAM than necessary.

There is the use case when the viz server needs to be hosted in an AWS EC2 t2.nano (0.5 GB) or t2.micro (1 GB).

Reply:
No, handle_step only does caching if you specify a specific step. If it is called without a step (like the BaseControls do) the model is simply advanced. Added a docstring to handle_step to clarify this. There is still the model cache and to keep the code simple it always holds the initial step. But I think we can live with 1 additional copy. That might also be handy if someone wants to add the option to go to the first step later, to re-run the model. Currently it is a bit ambiguous if the reset button should just go back to the first step or regenerate a new model (It does the latter but there is no reason why it shouldn't just rewind).

Reply:
But in the hypothetical situation when I am not familiar with the frontend code, and want to read a small subset of this component, I have to read the docstring about caching and so on. Whereas if you have a simpler `handle_step` for the simpler controller, I can grasp the structure better before moving on to a more complicated component.

Reply:
As in, have 2 `handle_step`-like functions for each controller classes.

Reply:
> But in the hypothetical situation when I am not familiar with the frontend code, and want to read a small subset of this component, I have to read the docstring about caching and so on. 

That doesn't make sense. If you want to understand the code, why wouldn't you read the docstring? 
 And the whole function is only 10 lines of code, 5 of which would need to be duplicated. I think this would be more confusing and more difficult to maintain. And imagine someone writing a new Controls component. Now they have to decide which handle step method they want to use or add yet another one. If the function were more complex I would agree but I think it is not worth the complications.

Is this a blocker?

Reply:
> Now they have to decide which handle step method they want to use or add yet another one.

This is far better than fattening up `handle_step` with more if-else's if the existing function does not cover the use case.

Why do I need to go through the caching code and have to consciously read and then to ignore it, just to read about stepping for basic controls? Such a waste of effort. I strongly disagree.

Reply:
I think we need a third opinion. My final statement is that my experience as a software developer tells me that in 95% of cases one should aim for maintainability. And two functions are harder to maintain than one.